### ISSUE TITLE
Integration of new configs for Test Flight v2

### ISSUE THREAD

**Issue by jwvanderbeck** [KSP-RO/RealismOverhaul#2315]
I have been wanting to release Test Flight v2 for a while now, but I am unsure how I can properly do so given dependencies on RO and RP1 that are further complicated by TestLite which stole the old Test Flight config structure and most likely won't work with the new one without updates on the TL side.

So tl;dr if RO/RP1 TF configs get updates, it will break TL.  So now TF is held up because of TL.  Not sure what to do.

Open to suggestions here.

If I release TF v2 as is, then I will need to continue to also include modified patches that the user has to manually copy into RO/RP1.  CKAN installs would be a mess because they would not install the extra patches and thus would not function at all.  This is why TF v2 is still marked "pre-release", so that CKAN doesn't pick it up.

[TestFlight2-RO_RP1-Patch.zip](https://github.com/KSP-RO/RealismOverhaul/files/4870868/TestFlight2-RO_RP1-Patch.zip)

**Comment by DRVeyl** [comment]
Sneak back to the Discord, chat with myself and @ec429.  Needs to clarify if TL is stealing the original declarations, or if it is taking from the output of that MM-fu that we redesigned for TF v2.

**Comment by jwvanderbeck** [comment]
Good point, it might be stealing the output in which case it **may** just work.

**Comment by ec429** [comment]
TestLite takes the `TESTFLIGHT` nodes and processes them with some MM-fu that's basically TFv1's MM with all the bits I didn't need ripped out.

Is there some documentation of what you've changed?  We may be able to do some MM magic that generates the right thing based on which of us is installed.

**Comment by jwvanderbeck** [comment]
> TestLite takes the `TESTFLIGHT` nodes and processes them with some MM-fu that's basically TFv1's MM with all the bits I didn't need ripped out.
> 
> Is there some documentation of what you've changed? We may be able to do some MM magic that generates the right thing based on which of us is installed.

This is the final (post-patch) config:
https://github.com/KSP-RO/TestFlight/wiki/New-Config-Style

Essentially the content is the same, but it now works similar to how RF works, with a single module per part and multiple CONFIGs in that module, rather than the old method of having one full set of modules per CONFIG.

As for the intermediate configs, I'm not 100% sure what changed there as I don't understand MM much.  @DRVeyl graciously handled that, and I attached the new setup in the original post.

**Comment by DRVeyl** [comment]
All of the new configs for TF currently have :NEEDS[!TestLite,TestFlight].  So if TL is installed, none of the new pre-processing should take place, and TL's MM-fu should run as normal.  The original TESTFLIGHT{} nodes are created as they always have been.

TL might benefit from some of the newer preprocessing patches for TF, ie it might want to update its stealTF.cfg and some of the :NEEDS guards could get changed/removed if it will help.  But I think no changes are needed, TestFlight can release as-is, and TL would still work if RO merged those new configs in.  ... Of course, I'm saying this just after some static analysis, you should actually try it. :)

**Comment by DRVeyl** [comment]
Did this get resolved?  There's probably some synchronization that might have to happen in terms of TF2 and RO releasing with the updated patches.

**Comment by jwvanderbeck** [comment]
Need someone to test the changes with TestLite and see that things
continue to function (or not).

- John Vanderbeck

**Comment by siimav** [comment]
I finally started a new career with the TF alpha and noticed that the du gain rates are suspiciously low. For example a full burn of WAC only took the engine from 2000 to 2053 du. Do you happen to know which changes might be responsible for substantially reducing the data gain rate?
Btw, would be nice to have you around on the KSP-RO discord as well. :)

**Comment by siimav** [comment]
Did some MM config cache comparisons and found a significant dfference in `FlightDataRecorder_Engine` module. In the new configs `flightDataMultiplier` field is always `1.0` wheras in the old one WAC had the following: `flightDataMultiplier = 13.6170212765957`.

**Comment by jwvanderbeck-Unity** [comment]
@DRVeyl reworked the MM patches to account for the changes in TF V2 where all the configs for a given engine are in the same module now.

Is it possible that something to broken which is causing the change you see?

**Comment by DRVeyl** [comment]
I did, @siimav tested the fix.  Hadn't pushed it back here yet.
[TestFlight_Generic_Engines.cfg.txt](https://github.com/KSP-RO/RealismOverhaul/files/5010920/TestFlight_Generic_Engines.cfg.txt)

**Comment by jwvanderbeck** [comment]
Thanks I'll update the files on the TF side.

**Comment by jwvanderbeck** [comment]
Updated in https://github.com/KSP-RO/TestFlight/releases/tag/2.0.0-Alpha.7

**Comment by siimav** [comment]
Now that TF is properly following tech transfer limits and the research feature is disabled, I would suggest increasing the tech transfer cap from 2000 to 2500. The issue is that with some configs there would be a substantial decrease in reliability when transitioning from an older config to newer one. AJ10-42 -> AJ10-101A is a great example here. With -42 I have almost max data which means that cycle and ignition reliability are in high 90s. Carrying 2000 data over to -101A would only give me low to mid 90s reliability. Because of this I would be better off using the older and ever so slightly worse-performing configs rather than upgrading to the newest.
Because of the kink inside the curves, I think increasing the transfer to 2500 would already improve the situation by a lot.

**Comment by ec429** [comment]
> Because of this I would be better off using the older and ever so slightly worse-performing configs rather than upgrading to the newest.

Isn't that intended behaviour, though?  That engine upgrades should get tested out with low-value payloads, while important things like planetary probes sometimes use an older but more proven config?

**Comment by siimav** [comment]
I've been running the TF alpha for my entire career and I'd say it's ready for an official release.

**Comment by Capkirk123** [comment]
I have also been using TF v2.0 alpha for several months and it works great.

**Comment by jwvanderbeck** [comment]
Has anyone tested using the new patches but with TestLite to ensure functionality remains unchanged?

**Comment by lpgagnon** [comment]
I switched back to TestLite a few days ago after trying out TF2.0; everything seems to be working fine with the new patches

**Comment by jwvanderbeck** [comment]
In that case I would say we're good to go?  Once you all add the new configs to RO/RP1 and do a release, let me know and I'll cut a full release of TF v2

**Comment by DRVeyl** [comment]
3 months seems a good revisit rate for this.  There doesn't appear to be a PR in RO, so I'll need to commit the latest version of the patches directly, I guess.  Looks like it is what is above in my 1 Aug comment, or wrapped in the current TF pre-rel.

Agathorn, can you go ahead and release TF v2?  Since you built it to work with the previous configs, I believe that needs to happen first.  I cannot introduce the new config style into RO while users only have TF v1.

**Comment by jwvanderbeck** [comment]
The release shuffle is always a pain.  I can't release v2 until the configs
are in RO, but if you release RO with the new configs before v2 then TF
breaks :)

The best thing to do is get the releases ready to go so they are just a
button press away then do them at the same time.  TF v2 just needs to be
re-released without the pre-release tag (and without the configs) so that
CKAN can see it.

- John Vanderbeck
- http://www.johnvanderbeck.com
